apps/server/src/api-data/db/migration/db.migration.v3.ts

```diff
@@ -0,0 +1,410 @@
+import {
+  AutomationSettings,
+  CustomFields,
+  DatabaseModel,
+  EndAction,
+  EntryCustomFields,
+  NormalisedAutomation,
+  OntimeEntry,
+  ProjectData,
+  ProjectRundowns,
+  Rundown,
+  Settings,
+  SupportedEntry,
+  TimerLifeCycle,
+  Trigger,
+  URLPreset,
+  ViewSettings,
+} from 'ontime-types';
+import { is } from '../../../utils/is.js';
+import { dbModel } from '../../../models/dataModel.js';
+import { customFieldLabelToKey, checkRegex, isKnownTimerType, validateEndAction } from 'ontime-utils';
+import { event as eventModel } from '../../../models/eventsDefinition.js';
+
+// the methodology of the migrations is to just change the necessary keys to match with v4
+// and then let the normal project parser handle ensuring the the file is correct
+// we should also avoid relying on the types package as this file should continue to work with old types when things change
+
+export function shouldUseThisMigration(jsonData: object): boolean {
+  return (
+    is.objectWithKeys(jsonData, ['settings']) &&
+    is.object(jsonData.settings) &&
+    is.objectWithKeys(jsonData.settings, ['version']) &&
+    typeof jsonData.settings.version === 'string' &&
+    jsonData.settings.version.split('.')[0] === '3'
+  );
+}
+
+export function migrateAllData(jsonData: object) {
+  const settings = migrateSettings(jsonData);
+  const viewSettings = migrateViewSettings(jsonData);
+  const urlPresets = migrateURLPresets(jsonData);
+  const project = migrateProjectData(jsonData);
+  const migratedCustom = migrateCustomFields(jsonData);
+  const customFields = migratedCustom?.customFields;
+  const rundowns = migrateRundown(jsonData, migratedCustom?.translationTable);
+  const automation = migrateAutomations(jsonData);
+
+  return { ...jsonData, settings, viewSettings, urlPresets, project, customFields, rundowns, automation };
+}
+
+type old_Settings = {
+  version: string;
+  serverPort: number;
+  editorKey: null | string;
+  operatorKey: null | string;
+  timeFormat: '12' | '24';
+  language: string;
+};
+
+/**
+ * migrates a settings from v3 to v4.0.0
+ * - update the version number
+ */
+export function migrateSettings(jsonData: object): Settings | undefined {
+  if (is.objectWithKeys(jsonData, ['settings']) && is.object(jsonData.settings)) {
+    const oldSettings = structuredClone(jsonData.settings) as old_Settings;
+
+    const migrated: Settings = { ...oldSettings, version: '4.0.0' };
+    return migrated;
+  }
+}
+
+type old_ViewSettings = {
```

Comment:
question: could we avoid exporting these types?

Reply:
they are not exported..?